zipkin-server/src/main/java/zipkin2/server/internal/ZipkinHttpCollector.java

```diff
@@ -106,10 +106,15 @@ public HttpResponse uploadSpansV1Thrift(ServiceRequestContext ctx, HttpRequest r
   }
 
   /** This synchronously decodes the message so that users can see data errors. */
-  HttpResponse validateAndStoreSpans(SpanBytesDecoder decoder, ServiceRequestContext ctx, HttpRequest req) {
+  HttpResponse validateAndStoreSpans(SpanBytesDecoder decoder, ServiceRequestContext ctx,
+    HttpRequest req) {
     CompletableCallback result = new CompletableCallback();
 
-    req.aggregateWithPooledObjects(ctx.eventLoop(), ctx.alloc()).handle((msg, t) -> {
+    CompletableFuture<AggregatedHttpRequest> requestFuture = ctx.makeContextAware(
```

Comment:
this, I believe, was the key

Reply:
Hmm - I'm a bit surprised that the aggregator doesn't call the callback on the right executor, but given it seems to be the case, I think you can make it simpler with `req.aggregate(...).handleAsync((), ctx.contextAwareExecutor())`